### DISCUSSION TITLE
Raspberry Pi Pico: nRF24L01 over SPI

### DISCUSSION THREAD

**Discussion by melvincramer911** [micropython/micropython#10155] [Libraries & Drivers]
Hello 

### How do I get my required library for my pico, hopefully using thonny to get the library ? 

I have two pico's  two nrf24l01's and two push switches. 

I am trying to use this example and download the drivers ( I use thonny ), when I click on the link I to get the library, I get error message:  The 'micropython/micropython' repository doesn't contain the 'drivers/nrf24l01' path in 'master'.  

Here is my system: 
If I use Thonny [ tools, Mange packages ] and try to install anything with nrf24 in it I get error message:   this does not look like a Micropython/circuitpython package, are you sure you want to install it ? 

    
**### Here is the code:**  

from machine import Pin, SPI
import struct

from nrf24l01 import NRF24L01

led = Pin(25, Pin.OUT)                # LED
btn = Pin(28, Pin.IN, Pin.PULL_DOWN)  # button press
csn = Pin(15, mode=Pin.OUT, value=1)  # chip select not
ce  = Pin(14, mode=Pin.OUT, value=0)  # chip enable

# Addresses are in little-endian format. They correspond to big-endian
# 0xf0f0f0f0e1, 0xf0f0f0f0d2 - swap these on the other Pico!
pipes = (b"\xe1\xf0\xf0\xf0\xf0", b"\xd2\xf0\xf0\xf0\xf0")

def setup():
    nrf = NRF24L01(SPI(0), csn, ce, payload_size=4)
    
    nrf.open_tx_pipe(pipes[0])
    nrf.open_rx_pipe(1, pipes[1])
    nrf.start_listening()

    led.value(0)
    return nrf

def demo(nrf):
    state = 0 
    while True:
        if state != btn.value():
            state = btn.value()
            led.value(state)
            
            print("tx", state)
            nrf.stop_listening()
            try:
                nrf.send(struct.pack("i", state))
            except OSError:
                print('message lost')
            nrf.start_listening()
            
        if nrf.any():
            buf = nrf.recv()
            got = struct.unpack("i", buf)[0]
            print("rx", got)
            led.value(got)

def auto_ack(nrf):
    nrf.reg_write(0x01, 0b11111000)  # enable auto-ack on all pipes
    

nrf = setup()
auto_ack(nrf)
demo(nrf)

**Comment by jimmo** [comment]
@melvincramer911 

I'm not sure about Thonny, but the driver you're looking for is here: https://github.com/micropython/micropython-lib/tree/master/micropython/drivers/radio/nrf24l01

**Reply by peterhinch** [reply]
Indeed. My approach would be to use that driver and run `nrf24l01test.py` on both Picos. Run `initiator` on one and `responder` on the other. If you wire the radios as per the test script it will work. Then start adapting the script for your purposes.

**Reply by melvincramer911** [reply]
Hello Jimmo and Peter

I am pretty sure this is exactly what I needed.  Thank you. 

Sounds like you know this module and it is just what I need, my #1 chip, but I always hear bad things about this chip.  Where do you get your chips that work?  I am an amazon person myself.   I know that if you use the antenna, you need the adapter that sends more power and all systems seem to need a capacitor to the vcc.  

I am playing with a garden bot and I really appreciate this.    

Regards

David

**Reply by peterhinch** [reply]
> Where do you get your chips that work?

A lot of NRF24L01 modules on eBay are rubbish. Sparkfun ones always work well in my experience.

**Comment by melvincramer911** [comment]
Hello Peter, I have been going through this for ... a month .... old links send me to a non-exhistant library Can I buy you a couple coffees ?  Thank you soo much Oh and I have a dream of this pcb:has holes to solder in picohas nrf module included ( one that works ).  You can make a deal with someone that sells the stuff do you do not have to touch it. has solder holes for nrfcompletes the board out with generic solder holes, you know the ones that have 5 holes in parallel. Thank you very much Peter !!! > -------- Original Message --------> Subject: Re: [micropython/micropython] Raspberry Pi Pico: nRF24L01 over> SPI (Discussion #10155)> From: Peter Hinch ***@***.***>> Date: Tue, December 13, 2022 8:26 am> To: micropython/micropython ***@***.***>> Cc: melvincramer911 ***@***.***>,  Mention> ***@***.***>> > > Indeed. My approach would be to use that driver and run `nrf24l01test.py` on both Picos. Run `initiator` on one and `responder` on the other. If you wire the radios as per the test script it will work. Then start adapting the script for your purposes.> > -- > Reply to this email directly or view it on GitHub:> https://github.com/orgs/micropython/discussions/10155#discussioncomment-4390059> You are receiving this because you were mentioned.> > Message ID: ***@***.***>

**Comment by melvincramer911** [comment]
Hello Peter

I must be missing something

the NRF modules I see there are like 22 dollars and 11 bucks for the
antenna.

I was thinking about buying a dozen of them so like 400$ plus shipping.

looks like I will need to get a second job to support my hobby

Regards

David

**Comment by peterhinch** [comment]
Options are to see if your supplier will give a quantity discount. Or buy a few cheap ones from different sources, find one that works, then buy some more.

Maybe I was unlucky, but the cheap ones I bought could barely communicate across my very small office/lab.

**Comment by melvincramer911** [comment]
Thank you Peter

Google says you are correct, I say you are a great guy and it is a shame
there is not a pico - nrf !!!

Thank you again

**Comment by jouellnyc** [comment]
Hello. I came across this post as I was trying to get these modules installed myself. I bought about six of them from AliExpress.  Cheap, some $1, some 50 cents US. Having not used these I always keep risk low the first time.  :)

It turns out that some of them were definitely "counterfeit" and it appears that some were "legitimate". 

I'm using those terms loosely --  that is to say some of them worked and some of them straight up didn't. 

The ones that did work  had  mapping of the gpios on them:

<img width="320" height="426" alt="image" src="https://github.com/user-attachments/assets/9dcd3857-6e33-4ec6-8590-af6879ef0bb0" />

The ones that did not were just blank:

<img width="320" height="426" alt="image" src="https://github.com/user-attachments/assets/d00b239c-ebd1-43fc-ad93-b9ab507850bf" />


Only problem is each listing on the site showed them as blank and I don't see a way to order 'the good ones'. 

That being the case, when  I tried to use the ones that were working I only  ot  about 25% success transmission at first -- at about 12-18 inches distance. With help of Claude AI,  I was able to come up with some optimizations that got them  working to a place of about 75%-85% successful transmissions 

I hope this information helps. 

The code and the optimizations are included below -- it's basically the modified test file that comes w/the repo.

Mostly these:

```
 nrf.reg_write(0x06, 0x27)  # RF_SETUP: 250kbps, 0dBm power 
 nrf.reg_write(0x01, 0x00)  # Disable auto-ack on all pipes
 nrf.reg_write(0x04, 0xFF)  # SETUP_RETR: max delay, 15 retries
```

and some of the initial delays. Yes, the product is now weakened version BUT I get to play w/the toy now :).

AND I add tried a 10 µF and 47 µF capacitor between VCC/GND on each module.  47 µF seemed to work best.

Just my experience - I have about 24 hours worth of 'expertise',  so grain of salt here.

Best Regards


```
"""Test for nrf24l01 module.  Portable between MicroPython targets."""

import sys
import struct
import utime
from machine import Pin, SPI, SoftSPI
#from nrf24l01 import NRF24L01
from .nrf24l01 import NRF24L01
from micropython import const

# Responder pause between receiving data and checking for further packets.
_RX_POLL_DELAY = const(100)
# Responder pauses an additional _RESPONER_SEND_DELAY ms after receiving data and before
# transmitting to allow the (remote) initiator time to get into receive mode. The
# initiator may be a slow device. Value tested with Pyboard, ESP32 and ESP8266.
_RESPONDER_SEND_DELAY = const(100)

if sys.platform == "pyboard":
    spi = SPI(2)  # miso : Y7, mosi : Y8, sck : Y6
    cfg = {"spi": spi, "csn": "Y5", "ce": "Y4"}
elif sys.platform == "esp8266":  # Hardware SPI
    spi = SPI(1)  # miso : 12, mosi : 13, sck : 14
    cfg = {"spi": spi, "csn": 4, "ce": 5}
elif sys.platform == "esp32":  # Software SPI
    spi = SoftSPI(sck=Pin(25), mosi=Pin(33), miso=Pin(32))
    cfg = {"spi": spi, "csn": 26, "ce": 27}
elif sys.platform == "rp2":  # Hardware SPI with explicit pin definitions
    spi = SPI(0, sck=Pin(2), mosi=Pin(3), miso=Pin(4))
    cfg = {"spi": spi, "csn": 5, "ce": 6}
else:
    raise ValueError("Unsupported platform {}".format(sys.platform))


# Addresses are in little-endian format. They correspond to big-endian
# 0xf0f0f0f0e1, 0xf0f0f0f0d2
pipes = (b"\xe1\xf0\xf0\xf0\xf0", b"\xd2\xf0\xf0\xf0\xf0")


def initiator():
    csn = Pin(cfg["csn"], mode=Pin.OUT, value=1)
    ce = Pin(cfg["ce"], mode=Pin.OUT, value=0)
    spi = cfg["spi"]
    nrf = NRF24L01(spi, csn, ce, payload_size=8)
    
    # Back to your best working settings
    nrf.reg_write(0x06, 0x27)  # RF_SETUP: 250kbps, 0dBm power 
    nrf.reg_write(0x01, 0x00)  # Disable auto-ack on all pipes
    nrf.reg_write(0x04, 0xFF)  # SETUP_RETR: max delay, 15 retries
    
    print("NRF24L01 detected:", nrf.reg_read(0x10) == 0x73)

    nrf.open_tx_pipe(pipes[0])
    nrf.open_rx_pipe(1, pipes[1])
    nrf.start_listening()

    num_needed = 16
    num_successes = 0
    num_failures = 0
    led_state = 0

    print("NRF24L01 initiator mode, sending %d packets..." % num_needed)

    while num_successes < num_needed and num_failures < num_needed:
        # stop listening and send packet
        nrf.stop_listening()
        utime.sleep_ms(2)  # Small delay to ensure mode switch completes
        
        millis = utime.ticks_ms()
        led_state = max(1, (led_state << 1) & 0x0F)
        print("sending:", millis, led_state)
        try:
            nrf.send(struct.pack("ii", millis, led_state))
        except OSError:
            pass

        # start listening again
        nrf.start_listening()
        utime.sleep_ms(5)  # Small delay to ensure mode switch completes

        # wait for response, with 500ms timeout
        start_time = utime.ticks_ms()
        timeout = False
        while not nrf.any() and not timeout:
            if utime.ticks_diff(utime.ticks_ms(), start_time) > 500:
                timeout = True

        if timeout:
            print("failed, response timed out")
            print("NRF status:", hex(nrf.reg_read(0x07)))  # STATUS register
            print("RX FIFO empty:", not nrf.any())
            num_failures += 1
        else:
            # recv packet
            (got_millis,) = struct.unpack("i", nrf.recv())

            # print response and round-trip delay
            print(
                "got response:",
                got_millis,
                "(delay",
                utime.ticks_diff(utime.ticks_ms(), got_millis),
                "ms)",
            )
            num_successes += 1

        # delay then loop
        utime.sleep_ms(250)

    print("initiator finished sending; successes=%d, failures=%d" % (num_successes, num_failures))


def responder():
    csn = Pin(cfg["csn"], mode=Pin.OUT, value=1)
    ce = Pin(cfg["ce"], mode=Pin.OUT, value=0)
    spi = cfg["spi"]
    nrf = NRF24L01(spi, csn, ce, payload_size=8)
    
    # Back to your best working settings
    nrf.reg_write(0x06, 0x27)  # RF_SETUP: 250kbps, 0dBm power 
    nrf.reg_write(0x01, 0x00)  # Disable auto-ack on all pipes
    nrf.reg_write(0x04, 0xFF)  # SETUP_RETR: max delay, 15 retries
    
    print("NRF24L01 detected:", nrf.reg_read(0x10) == 0x73)
    
    nrf.open_tx_pipe(pipes[1])
    nrf.open_rx_pipe(1, pipes[0])
    nrf.start_listening()

    print("NRF24L01 responder mode, waiting for packets... (ctrl-C to stop)")

    while True:
        if nrf.any():
            while nrf.any():
                buf = nrf.recv()
                millis, led_state = struct.unpack("ii", buf)
                print("received:", millis, led_state)
                for led in leds:
                    if led_state & 1:
                        led.on()
                    else:
                        led.off()
                    led_state >>= 1
                utime.sleep_ms(_RX_POLL_DELAY)

            # Give initiator time to get into receive mode.
            utime.sleep_ms(_RESPONDER_SEND_DELAY)
            nrf.stop_listening()
            utime.sleep_ms(2)  # Small delay to ensure mode switch completes
            
            # Send response (only once, with error handling)
            try:
                print(f"About to send response: {millis}")
                nrf.send(struct.pack("i", millis))
                print("Send completed successfully")
            except OSError as e:
                print(f"Send failed: {e}")
            
            print("sent response")
            nrf.start_listening()


try:
    import pyb
    leds = [pyb.LED(i + 1) for i in range(4)]
except:
    leds = []

print("NRF24L01 test module loaded")
print("NRF24L01 pinout for test:")
print("    CE on", cfg["ce"])
print("    CSN on", cfg["csn"])
print("    SPI on", cfg["spi"])
print("run nrf24l01test.responder() on responder, then nrf24l01test.initiator() on initiator")
```
